### PR TITLE
Better state and disease parsing

### PR DESCRIPTION
**Pull request by natemcintosh**
## Nate's Summary
This adds the ability for the config generator to not only take in `"all"` or a single state/disease, but also a comma-separated list of states or diseases.

Added tests that cover all the possible cases (I think). If you think of a missing case, please let me know.

Finally, also made some small changes to other tests to make the type checker happy.

## Copilot's Summary
This pull request refactors the argument validation logic in `functions.py` by introducing two new helper functions, `parse_state` and `parse_disease`, to improve modularity and maintainability. Additionally, new test cases are added to validate these functions, and existing test cases are updated to reflect changes in the argument structure.

### Refactoring and modularization:
* Introduced `parse_state` and `parse_disease` functions to handle parsing and validation of `state` and `disease` arguments, replacing inline logic in `validate_args`. These functions support single values, comma-separated lists, and the keyword "all" while raising appropriate errors for invalid inputs. (`src/cfa_config_generator/utils/epinow2/functions.py`) [[1]](diffhunk://#diff-ab59143b1ba343d18c0ed85e6e50cacaa76d64fad93c142b8f349b5e61db04f8L232-R235) [[2]](diffhunk://#diff-ab59143b1ba343d18c0ed85e6e50cacaa76d64fad93c142b8f349b5e61db04f8R399-R486)

### Test enhancements:
* Updated existing test cases (`test_validate_args_default`, `test_invalid_state`, `test_invalid_disease`, etc.) to reflect the new structure, including the addition of a default `data_container` value. (`tests/test_args.py`) [[1]](diffhunk://#diff-d5d888969fd8e06ca26b226838d662a18b517912f4df62b41e87e8f8a2e1d9ffL78-R80) [[2]](diffhunk://#diff-d5d888969fd8e06ca26b226838d662a18b517912f4df62b41e87e8f8a2e1d9ffL90-R92) [[3]](diffhunk://#diff-d5d888969fd8e06ca26b226838d662a18b517912f4df62b41e87e8f8a2e1d9ffL108-R110) [[4]](diffhunk://#diff-d5d888969fd8e06ca26b226838d662a18b517912f4df62b41e87e8f8a2e1d9ffL127-R129) [[5]](diffhunk://#diff-d5d888969fd8e06ca26b226838d662a18b517912f4df62b41e87e8f8a2e1d9ffL152-R154) [[6]](diffhunk://#diff-d5d888969fd8e06ca26b226838d662a18b517912f4df62b41e87e8f8a2e1d9ffL174-R266)
* Added parameterized test cases `test_state_parsing` and `test_disease_parsing` to thoroughly test the new `parse_state` and `parse_disease` functions, covering valid inputs, edge cases, and failure scenarios. (`tests/test_args.py`)

### Codebase improvements:
* Imported `parse_state` and `parse_disease` into the test suite to ensure comprehensive testing. (`tests/test_args.py`)

### CODE REVIEW

File: src/cfa_config_generator/utils/epinow2/functions.py [hunk 1]
```diff
@@ -407,3 +396,91 @@ def exclude_task(config_data: list[dict], filters: dict[str, list[str]]) -> list
     ]
 
     return filtered_data
+
+
+def parse_state(state: str) -> list[str]:
```
Comment (athowes) [inline src/cfa_config_generator/utils/epinow2/functions.py:401]: Feels likely that parsing single values, CSV values, and all into a list could be abstracted away from relating to states or diseases? (Essentially a suggestion/question on DRYing this.)

Comment (athowes) [reply src/cfa_config_generator/utils/epinow2/functions.py:401]: This would also help with what looks like test duplication

File: src/cfa_config_generator/utils/epinow2/functions.py [hunk 2]
```diff
@@ -407,3 +396,91 @@ def exclude_task(config_data: list[dict], filters: dict[str, list[str]]) -> list
...
+    """
+    Parses the state argument and returns a list of states.
+
+    Parameters:
+    ----------
+        state: str
```
Comment (athowes) [inline src/cfa_config_generator/utils/epinow2/functions.py:407]: Would it be reasonable to also allow users to pass in the list (i.e. not only a string)?

**Comment by athowes** [comment]
My main comment is about the generality. I can imagine we might want to be able to parse things other than states and diseases as well, so the effort to abstract a parser out could be more worth it. Up to your discretion.

File: src/cfa_config_generator/utils/epinow2/functions.py (hunk 2 above)
Comment (natemcintosh) [reply src/cfa_config_generator/utils/epinow2/functions.py:407]: That's a good question. The `validate_args()` function is where we parse and validate the input from the command line, which gets the states and diseases as a single string.

In theory, we could have the command line parser itself create a list from the input, but this felt simpler.

File: src/cfa_config_generator/utils/epinow2/functions.py (hunk 1 above)
Comment (natemcintosh) [reply src/cfa_config_generator/utils/epinow2/functions.py]: Yea, I like this idea. Swapped it to a more generic function that can take any list of valid options.

File: tests/test_args.py [hunk 3]
```diff
@@ -171,10 +168,66 @@ def test_invalid_disease_exclusion():
...
+        ("OH", nssp_valid_states, False),
+        ("CA", nssp_valid_states, False),
+        ("WA ", nssp_valid_states, False),
+        (" WA", nssp_valid_states, False),
+        (" WA ", nssp_valid_states, False),
+        # Test failures,
+        ("ZZ,COVID-19", all_diseases, True),
+        ("COVID-19,ZZ", all_diseases, True),
+        ("ZZ,WA", nssp_valid_states, True),
+        ("OH,WA,OO", nssp_valid_states, True),
```
Comment (kgostic) [inline tests/test_args.py:210]:
```suggestion
        ("OH,WA,OO", nssp_valid_states, True),
        (32, nssp_valid_states, True),
        (["OH","WA","OO"], nssp_valid_states, True)
```

Add some weird types to the call, here a number and a list?

File: src/cfa_config_generator/utils/epinow2/functions.py [hunk 4]
```diff
@@ -407,3 +398,56 @@ def exclude_task(config_data: list[dict], filters: dict[str, list[str]]) -> list
...
+                        " Valid options are {valid_options}."
+                    )
+                )
+
+            return option_list
+        case _:
+            raise ValueError(
+                (
+                    f"Option {raw_input} not recognized. Valid options are 'all', "
+                    " a singleton, or a comma-separated list of options."
```
Comment (kgostic) [inline src/cfa_config_generator/utils/epinow2/functions.py:451]:
```suggestion
                    " a singleton, or a comma-separated list of options, passed as a string."
```

File: tests/test_args.py (hunk 3 above)
Comment (kgostic) [reply tests/test_args.py:210]: I think the list input is important to test here!

**Review by kgostic — APPROVED** [review]
This looks really nice!

I suggested a few small tweaks to address the possibility that someone enters a list instead of a string, which I think is likely to happen at some point. Otherwise I think this can merge.

**Comment by natemcintosh** [comment]
I figured we might as well add the option for parsing a list as well, since we're here in the PR.

**Review by athowes — APPROVED** [review]

**Comment by athowes** [comment]
Can we delete this branch?
